File: 1501~2000/1846. Maximum Element After Decreasing and Rearranging/main.py
"""
Source: https://leetcode.com/problems/maximum-element-after-decreasing-and-rearranging/
Date: 2023/4/12
Skill:
Ref:
Runtime: 505 ms, faster than 71.17%
Memory Usage: 24 MB, less than 67.48%
Time complexity:
Space complexity:
Constraints:
    1 <= arr[i] <= 10^9
"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools


class Solution:
    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:
        arr.sort()
        arr[0], sz = 1, len(arr)
        for i in range(1, sz):
            if arr[i] > arr[i - 1] + 1:
                arr[i] = arr[i - 1] + 1
        return arr[sz - 1]


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1846. Maximum Element After Decreasing and Rearranging/main.cpp
/**
 * Source: ibit.ly/fyGsG
 * Date: 2023/11/15
 * Skill:
 * Runtime: 76 ms, faster than 69.40% of C++ online submissions
 * Memory Usage: 51.52 MB, less than 56.72% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    int maximumElementAfterDecrementingAndRearranging(vector<int>& arr) {
        sort(begin(arr), end(arr));
        arr[0] = 1;
        for (int i = 1; i < arr.size(); i++) {
            if (arr[i] - arr[i - 1] > 1) {
                arr[i] = arr[i - 1] + 1;
            }
        }
        return arr.back();
    }
};


int main() {
    Solution s;
}